### PR TITLE
Add support to mount kubeconfig  in `tkn` task

### LINKED ISSUE #405: Tkn Task: Provide a way to mount credentials / kubeconfig
# Expected Behavior

The `tkn` task should be able to work with a kubeconfig for a cluster so that it can operate on it.

# Actual Behavior

At the moment there doesn't appear to be a good way to get a kubeconfig into the the Task.

Possible solutions to this problem would be:
- Cluster PipelineResource
- ["Optional" Workspace](https://github.com/tektoncd/pipeline/issues/2119) that would allow kubeconfig to be mounted
- A named volumeMount in the Step which a TaskRun could provide via podTemplate.volume

### PR DESCRIPTION
**Pull request by vinamra28**
# Changes

The tkn task will be able to work with a kubeconfig for a cluster so that it can operate on it. The following task can also accept a SCRIPT as input so that we can run multiple tkn commands.

Closes #405 

Signed-off-by: vinamra28 <vinjain@redhat.com>

/cc @sbwsg @vdemeester 

# Submitter Checklist

These are the criteria that every PR should meet, please check them off as you
review them:

- [X] Includes [docs](https://github.com/tektoncd/community/blob/master/standards.md#principles) (if user facing)
- [X] Commit messages follow [commit message best practices](https://github.com/tektoncd/community/blob/master/standards.md#commit-messages)
- [X] Complies with [Catalog Orgainization TEP][TEP], see [example]. **Note** [An issue has been filed to automate this validation][validation]
  - [X] File path follows  `<kind>/<name>/<version>/name.yaml`
  - [X] Has `README.md` at `<kind>/<name>/<version>/README.md`
  - [X] Has mandatory `metadata.labels` - `app.kubernetes.io/version` the same as the `<version>` of the resource
  - [X] Has mandatory `metadata.annotations` `tekton.dev/pipelines.minVersion`
  - [X] mandatory `spec.description` follows the convention

          ```

          spec:
            description: >-
              one line summary of the resource

              Paragraph(s) to describe the resource.
          ```

_See [the contribution guide](https://github.com/tektoncd/catalog/blob/master/CONTRIBUTING.md)
for more details._

---

[TEP]: https://github.com/tektoncd/community/blob/master/teps/0003-tekton-catalog-organization.md
[example]: https://github.com/tektoncd/catalog/tree/master/task/git-clone/0.1
[validation]:  https://github.com/tektoncd/catalog/issues/413

### CODE REVIEW

**Review by vdemeester — COMMENTED** [review]
SGTM, but I feel it would be even better if we had support for optional workspace wouldn't it ?
As of this PR, a `kubeconfig` configmap is required to run the task right ?

**Comment by vinamra28** [comment]
> SGTM, but I feel it would be even better if we had support for optional workspace wouldn't it ?
> As of this PR, a `kubeconfig` configmap is required to run the task right ?

nope. In the script I am checking for the `kubeconfig` file using if-condition. If the file is present then `export KUBECONFIG` else use the current cluster

**Comment by vdemeester** [comment]
> > SGTM, but I feel it would be even better if we had support for optional workspace wouldn't it ?
> > As of this PR, a `kubeconfig` configmap is required to run the task right ?
> 
> nope. In the script I am checking for the `kubeconfig` file using if-condition. If the file is present then `export KUBECONFIG` else use the current cluster

Right, but if the user doesn't provide a `workspace`, the `TaskRun` will fail because it is missing a workspace :sweat_smile:

**Comment by vinamra28** [comment]
> > > SGTM, but I feel it would be even better if we had support for optional workspace wouldn't it ?
> > > As of this PR, a `kubeconfig` configmap is required to run the task right ?
> > 
> > 
> > nope. In the script I am checking for the `kubeconfig` file using if-condition. If the file is present then `export KUBECONFIG` else use the current cluster
> 
> Right, but if the user doesn't provide a `workspace`, the `TaskRun` will fail because it is missing a workspace

ya true, or maybe an `emptyDir:{}` would work in that case
```yaml
apiVersion: tekton.dev/v1beta1
kind: TaskRun
metadata:
  name: tkn-run
spec:
  taskRef:
    name: tkn
  workspaces:
    - name: kubeconfig
      emptyDir: {}
  params:
    - name: SCRIPT
      value: |
        tkn task start $1
        tkn taskrun list
        tkn task logs $1 -f
    - name: ARGS
      value:
        - taskRunName
```
 But yes optional workspaces would be a good option :wink:

**Comment by vinamra28** [comment]
/hold
will wait for the `optional-workspaces` feature in `tektoncd/pipeline` :slightly_smiling_face:

**Comment by vinamra28** [comment]
Added support for optional workspaces
/cc @vdemeester @sbwsg 
/hold cancel

**Comment by vinamra28** [comment]
ohh...this needs a latest pipeline version bump having optional workspaces support :sweat_smile:

File: task/tkn/0.2/tkn.yaml [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+        An optional workspace into which a kubeconfig file called kubeconfig will
+        be written that will contain the information required to access the cluster.
+      optional: true
+  description: >-
+    This task performs operations on Tekton resources using tkn
+
+  params:
+    - name: TKN_IMAGE
+      description: tkn CLI container image to run this task
+      default: gcr.io/tekton-releases/dogfooding/tkn
```
Comment (vdemeester) [inline task/tkn/0.2/tkn.yaml:25]: we may want to use a tagged version of this image :panda_face:

**Comment by sbwsg** [comment]
Thanks @vinamra28 !

/approve

**Comment by vinamra28** [comment]
can we get this in?
/cc @vdemeester @chmouel

**Review by Divyansh42 — APPROVED** [review]
/lgtm

Thanks :+1:

**Comment by chmouel** [comment]
/lgtm
/approve

I think as a followup this can use a test even if it's using the same kubeconfig as in cluster, it will actually test the task, what do you think @vinamra28 ?

Thanks 👍🏼🔥

**Comment by vinamra28** [comment]
> /lgtm
> /approve
> 
> I think as a followup this can use a test even if it's using the same kubeconfig as in cluster, it will actually test the task, what do you think @vinamra28 ?
> 
> Thanks 👍🏼

Sure @chmouel. Probably in a followup PR I will surely try adding a test for the same :slightly_smiling_face:

**Pull request merged**